astlib/location.mli

```diff
@@ -0,0 +1,53 @@
+type t = Ocaml_common.Location.t = {
+  loc_start : Lexing.position;
+  loc_end : Lexing.position;
+  loc_ghost : bool;
+}
+
+type 'a loc = 'a Ocaml_common.Location.loc = { txt : 'a; loc : t }
+
+val none : t
+
+(** {1 Input info} *)
+
+val set_input_name : string -> unit
+
+module Location_error : sig
+  type location
+
+  type t
+
+  val is_well_formed : t -> bool
+
+  val main_msg : t -> string loc
+
+  val sub_msgs : t -> string loc list
+
+  val set_main_msg : t -> string -> t
+
+  val make : sub:string loc list -> string loc -> t
+
+  val update_loc : t -> location -> t
+
+  val of_exn : exn -> t option
+end
+with type location := t
+
+(** {1 Automatically reporting errors for raised exceptions} *)
+
+val register_error_of_exn : (exn -> Location_error.t option) -> unit
+(** Each compiler module which defines a custom type of exception which can
+    surface as a user-visible error should register a "printer" for this
+    exception using [register_error_of_exn]. The result of the printer is an
+    [error] value containing a location, a message, and optionally sub-messages
+    (each of them being located as well). *)
+
+val raise_errorf :
+  ?loc:t -> ('a, Format.formatter, unit, 'b) format4 -> 'a
+
+val report_exception : Format.formatter -> exn -> unit
+(** Reraise the exception if it is unknown. *)
+
+exception Error of Location_error.t
```

Comment:
Is there a better name we could come up for for this exception? `Error` is not very explicit. That way, I could also rename the `Location.Location_error` module simply to `Location.Error`.

Reply:
Not sure, this is used for various kinds of errors. Us there something preventing you from renaming `Location.Location_error` to `Location.Error` at the moment.

Reply:
I thought it might be confusing to have both an exception called `Error` and a module called `Error` inside `Location`. Do you think it's not?

Reply:
I think it's fine.